mesonbuild/scripts/meson_install.py

```diff
@@ -51,12 +51,25 @@ def __exit__(self, type, value, traceback):
         for d in self.dirs:
             append_to_log(d)
 
-def set_mode(path, mode):
-    if mode is None:
-        # Keep mode unchanged
+def is_executable(path):
+    # Check if any of the 'x' bits are set in the source file mode
+    return bool(os.stat(path).st_mode & 0o111)
+
+def sanitize_permissions(path, umask):
+    if umask is None:
         return
-    if (mode.perms_s or mode.owner or mode.group) is None:
-        # Nothing to set
+    new_perms = 0o777 if is_executable(path) else 0o666
+    new_perms &= ~umask
+    try:
+        os.chmod(path, new_perms)
+    except PermissionError as e:
+        msg = '{!r}: Unable to set permissions {!r}: {}, ignoring...'
+        print(msg.format(path, new_perms, e.strerror))
+
+def set_mode(path, mode, default_umask):
+    if mode is None or (mode.perms_s or mode.owner or mode.group) is None:
```

Comment:
I don't think this conditional is correct. E.g. `mode.perms_s == None` and `mode.owner == 0`.

Reply:
Indeed you are correct.

`mesonlib.FileMode(None, 0, None)` would fail this check, since the intended effect is to chown the file to root, but the check of (None or 0 or None) would produce None and the 0 would get buried in it...

But this is a pre-existing issue and fixing it actually affects the test cases... So I'd rather deal with this in a separate PR. Will do so.